File: yxb144325004祁陈斌/Notes/Notes/DrawView.m
//
//  DrawView.m
//  Notes
//
//  Created by xsdlr on 14/11/18.
//  Copyright (c) 2014年 xsdlr. All rights reserved.
//

#import "DrawView.h"
#import "PaintPath.h"

@interface DrawView()
@property (assign,nonatomic) CGMutablePathRef path;
@property (strong,nonatomic) NSMutableArray *pathArray;
@end

@implementation DrawView

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        _width = 15.0f;
        _color = [UIColor redColor];
        _path=CGPathCreateMutable();
        _pathArray = [NSMutableArray array];
    }
    return self;
}

- (void)drawRect:(CGRect)rect {
    CGContextRef context=UIGraphicsGetCurrentContext();
    for (PaintPath* path in _pathArray) {
        CGContextAddPath(context, path.path.CGPath);
        CGContextSetLineWidth(context, path.width);
        CGContextSetStrokeColorWithColor(context, [path.color CGColor]);
        CGContextSetLineCap(context, kCGLineCapRound);//圆形起笔
        CGContextDrawPath(context, kCGPathStroke);
    }
    
    if (_path) {
        CGContextAddPath(context, _path);
        CGContextSetLineWidth(context, _width);
        CGContextSetStrokeColorWithColor(context, [_color CGColor]);
        CGContextSetLineCap(context, kCGLineCapRound);
        CGContextDrawPath(context, kCGPathStroke);
    }
    
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    _path=CGPathCreateMutable();
    UITouch *touch=[touches anyObject];
    CGPoint location=[touch locationInView:self];
    CGPathMoveToPoint(_path, NULL, location.x, location.y);
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    UITouch *touch=[touches anyObject];
    CGPoint location=[touch locationInView:self];
    CGPathAddLineToPoint(_path, NULL, location.x, location.y);
    [self setNeedsDisplay];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    UIBezierPath* currentPath=[UIBezierPath bezierPathWithCGPath:_path];
    PaintPath* paintPath = [[PaintPath alloc] initWithColor:_color width:_width path:currentPath];
    [_pathArray addObject:paintPath];
    CGPathRelease(_path);
}

- (BOOL)clearView {
    [_pathArray removeAllObjects];
    [self setNeedsDisplay];
    return YES;
}

- (BOOL)saveToPNGFile:(NSString *)path {
    UIGraphicsBeginImageContext(self.bounds.size);
    [self.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    return [UIImagePNGRepresentation(image) writeToFile:path atomically:YES];
}
@end

File: 21451077闭鸿晖/HomeWork_Note/HomeWork_Note/DrawView.swift

import UIKit


class DrawView:UIView {
    //路径定义
    struct PathDef {
        //贝塞尔曲线
        var path:UIBezierPath
        var color:CGColorRef
    }
    //标示当前是否在绘制一条path
    var isDrawing=false
    //路径数组
    var paths:[PathDef]!=[]
    //当前从beging到end绘制的一条path
    var curPath:CGMutablePathRef!
    
    var curColor:CGColorRef!
    var color:Int{
        set{
            switch newValue{
            case 0:curColor=UIColor.blackColor().CGColor
            case 1:curColor=UIColor.redColor().CGColor
            case 2:curColor=UIColor.greenColor().CGColor
            case 3:curColor=UIColor.blueColor().CGColor
            case 4:curColor=UIColor.purpleColor().CGColor
            default:break
            }
        }
        get{
            return 0
        }
    }
    
    override func drawRect(rect: CGRect) {
        var context = UIGraphicsGetCurrentContext()
        func drawPath(#path:UIBezierPath,#color:CGColorRef){
            CGContextAddPath(context, path.CGPath)
            CGContextSetLineWidth(context, 1)
            CGContextSetStrokeColorWithColor(context,color)
            CGContextDrawPath(context, kCGPathStroke)
        }
        for myPath in paths{
            drawPath(path: myPath.path, color: myPath.color)
        }
        //当前正在drawing（touchesMoved阶段）
        if isDrawing{
            drawPath(path: UIBezierPath(CGPath: curPath), color: curColor)
        }
    
        
    }
    
    //MARK: - touchEvents
    override func touchesBegan(touches: NSSet, withEvent event: UIEvent) {
        isDrawing=true
        var touch=touches.anyObject() as UITouch
        var point=touch.locationInView(self)
        //创建一个可变路径
        curPath=CGPathCreateMutable()
        //移动到起点
        CGPathMoveToPoint(curPath, nil, point.x, point.y)
    }
    
    override func touchesMoved(touches: NSSet, withEvent event: UIEvent) { var touch=touches.anyObject() as UITouch
        var point=touch.locationInView(self)
        //添加直线到当前路径
        CGPathAddLineToPoint(curPath, nil, point.x, point.y)
        self.setNeedsDisplay()
    }
    
    override func touchesEnded(touches: NSSet, withEvent event: UIEvent) {
        var path:UIBezierPath=UIBezierPath(CGPath: curPath)
        paths.append(PathDef(path: path, color: curColor))
        isDrawing=false
    }
    
    
}
   